src/secrets/src/cache.rs

```diff
@@ -0,0 +1,762 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::BTreeMap;
+use std::fmt;
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use std::sync::{Arc, RwLock};
+use std::time::{Duration, Instant};
+
+use async_trait::async_trait;
+use mz_repr::GlobalId;
+
+use crate::{CachingSecretsController, SecretsController, SecretsReader};
+
+/// Default "time to live" for a single cache value, represented in __seconds__.
+pub const DEFAULT_TTL_SECS: AtomicU64 = AtomicU64::new(Duration::from_secs(300).as_secs());
+
+#[derive(Debug)]
+struct CachingParameters {
+    /// Whether caching is enabled, can be changed at runtime.
+    enabled: AtomicBool,
+    /// Cache values only live for so long.
+    ttl_secs: AtomicU64,
+}
+
+impl CachingParameters {
+    pub fn enabled(&self) -> bool {
+        self.enabled.load(Ordering::Relaxed)
+    }
+
+    pub fn set_enabled(&self, enabled: bool) -> bool {
+        self.enabled.swap(enabled, Ordering::Relaxed)
+    }
+
+    pub fn ttl(&self) -> Duration {
+        let secs = self.ttl_secs.load(Ordering::Relaxed);
+        Duration::from_secs(secs)
+    }
+
+    pub fn set_ttl(&self, ttl: Duration) -> Duration {
+        // Maintain a minimum TTL of 1 second just incase someone provides a Duration that is
+        // sub-second.
+        let secs = std::cmp::max(ttl.as_secs(), 1);
+        let prev = self.ttl_secs.swap(secs, Ordering::Relaxed);
+        Duration::from_secs(prev)
+    }
+}
+
+impl Default for CachingParameters {
+    fn default() -> Self {
+        CachingParameters {
+            enabled: AtomicBool::new(true),
+            ttl_secs: DEFAULT_TTL_SECS,
+        }
+    }
+}
+
+/// Values we store in the cache.
+///
+/// Note: we manually implement Debug to prevent leaking secrets in logs.
+struct CacheItem {
+    secret: Vec<u8>,
+    ts: Instant,
+}
+
+impl CacheItem {
+    pub fn new(secret: Vec<u8>, ts: Instant) -> Self {
+        CacheItem { secret, ts }
+    }
+}
+
+impl fmt::Debug for CacheItem {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("CacheItem")
+            .field("secret", &"( ... )")
+            .field("ts", &self.ts)
+            .finish()
+    }
+}
+
+#[derive(Debug)]
+pub struct BTreeCachingSecretsController<T> {
+    /// Our underlying secrets, source of truth.
+    inner: T,
+    /// In-memory cache.
+    ///
+    /// TODO(parkmycar): Consider using something like an LRU cache.
```

Comment:
I wouldn't worry about it! We have a very strict limit on `max_secrets` (100, I think?) that will nicely bound the size of this cache. Probably worth dropping that comment in here as justification and then you can avoid the TODO.

Reply:
Good point! I left a comment on the `CachingSecretsReader` that there is a limit in the number of secrets a user can create, thus we don't need to worry about a max size or eviction policy